blocks/library/search/index.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * WordPress dependencies
+ */
+import { __, _x } from '@wordpress/i18n';
+import { Component } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+ import './editor.scss';
+import { registerBlockType } from '../../api';
+import InspectorControls from '../../inspector-controls';
+import TextControl from '../../inspector-controls/text-control';
+
+class SearchBlock extends Component {
+	constructor() {
+		super( ...arguments );
+		this.updateLabel = this.updateLabel.bind( this );
+		this.updatePlaceholder = this.updatePlaceholder.bind( this );
+		this.updateSubmitValue = this.updateSubmitValue.bind( this );
+		this.handleFormSubmit = this.handleFormSubmit.bind( this );
+	}
+
+	updateLabel( newLabel ) {
+		this.props.setAttributes( { label: newLabel } );
+	}
+
+	updatePlaceholder( newPlaceholder ) {
+		this.props.setAttributes( { placeholder: newPlaceholder } );
+	}
+
+	updateSubmitValue( newValue ) {
+		this.props.setAttributes( { submitValue: newValue } );
+	}
+
+	handleFormSubmit( event ) {
+		event.preventDefault();
+		return false;
+	}
+
+	render() {
+		const {
+			attributes,
+			focus,
+		} = this.props;
+
+		const {
+			label,
+			placeholder,
+			submitValue,
+		} = attributes;
+
+		return [
+			<form
+				role="search"
+				method="get"
+				className="search-form"
+				onSubmit= { this.handleFormSubmit }
+			>
+				<label>
+					<span className="search-label">{ label }</span>
+					<input type="search" className="search-field" placeholder={ placeholder } value="" name="s" />
+				</label>
+				<input type="submit" className="search-submit" value={ submitValue } />
+			</form>,
+			focus && (
+				<InspectorControls key="inspector">
+					<TextControl label={ __( 'Search label' ) } value={ label } onChange={ this.updateLabel } help={ __( 'Provides a label for the search field.' ) } />
+					<TextControl label={ __( 'Placeholder text' ) } value={ placeholder } onChange={ this.updatePlaceholder } help={ __( 'Provides placeholder text for the search field.' ) } />
+					<TextControl label={ __( 'Submit text' ) } value={ submitValue } onChange={ this.updateSubmitValue } help={ __( 'The text for the form submit button.' ) } />
+				</InspectorControls>
+			),
+		];
+	}
+}
+
+const searchAttributes = {
+	label: {
+		type: 'string',
+		default: _x( 'Search for:', 'label' ),
+	},
+	placeholder: {
+		type: 'string',
+		default: _x( 'Search &hellip;', 'placeholder' ),
+	},
+	submitValue: {
+		type: 'string',
+		default: _x( 'Search', 'submit button' ),
+	},
```

Comment:
The strings here are duplicated here in JS and in PHP. It would be ideal if they were all consolidated in the one `register_block_type()` PHP call and then exported to JS.

Reply:
Yes, attributes should be defined on the server using PHP. It will solve the issues with code duplication. See Latest Posts block for reference:
https://github.com/WordPress/gutenberg/blob/e8f0f89c5a954d4d67c3bfac1cfa9ef0edd082ed/packages/block-library/src/latest-posts/index.php#L88-L130